### PR TITLE
Evaluate tuple elements in a switch control expression in the order written

### LINKED ISSUE #39564: Incorrect order of evaluation for tuple element expressions when the tuple is the operand of a switch expression
@atifaziz commented on [Tue Oct 29 2019](https://github.com/dotnet/docs/issues/15496)

I believe this belongs in the [**Switch expressions**](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#switch-expressions) section, but also in [the tutorial](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/pattern-matching).

There seems to be a subtle and historical change in how C# respects, or actually _no longer respects_, the _evaluation order_ when using pattern-matching in a switch-expression. This can bite people (as it bit me when refactoring code) in subtle ways if not documented and called out with a warning. At first I thought it was a bug, but then I think this subtle change is intended if I am reading the following [section](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/atterns#order-of-evaluation-in-pattern-matching) of the [C# 8 specification proposal on recursive pattern matching](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/patterns) right:

> ### Order of evaluation in pattern-matching
>
> Giving the compiler flexibility in reordering the operations executed during pattern-matching can permit flexibility that can be used to improve the efficiency of pattern-matching. The (unenforced) requirement would be that properties accessed in a pattern, and the Deconstruct methods, are required to be "pure" (side-effect free, idempotent, etc). That doesn't mean that we would add purity as a language concept, only that we would allow the compiler flexibility in reordering operations.
>
> **Resolution 2018-04-04 LDM**: confirmed: the compiler is permitted to reorder calls to `Deconstruct`, property accesses, and invocations of methods in `ITuple`, and may assume that returned values are the same from multiple calls. The compiler should not invoke functions that cannot affect the result, and we will be very careful before making any changes to the compiler-generated order of evaluation in the future.

To demonstrate the problem in action, suppose the following code:

```c#
var s = new[] { "foo", "bar" };
var i = 0;
string Foobar() => s[i++];
var foobar = (Foobar(), Foobar());
var r = foobar switch
{
    ("foo", "bar") =>  "Yep, all good!",
    var (a, b)     => $"Wait, what? I got ({a}, {b})!"
};
Console.WriteLine(r);
```

The code prints the string from the first matching case, as one would expect:

    Yep, all good!

But if one substitutes the `foobar` variable use directly with the right-hand expression of its variable declaration assignment, like so:

```c#
var s = new[] { "foo", "bar" };
var i = 0;
string Foobar() => s[i++];
// var foobar = (Foobar(), Foobar());
var r = /*foobar*/ (Foobar(), Foobar()) switch
{
    ("foo", "bar") =>  "Yep, all good!",
    var (a, b)     => $"Wait, what? I got ({a}, {b})!"
};
Console.WriteLine(r);
```

The output instead changes to:

    Wait, what? I got (bar, foo)!

It seems as if the compiler actually switches around the invocations of the local function `Foobar` and the side-effects manifest as a bug! I used the [ILSpy](http://www.ilspy.net/) decompiler to look at the compiled code and one can see what's effectively going on (comment added by me):

```c#
string[] s = new string[2]
{
    "foo",
    "bar"
};
int i = 0;
string b = Foobar(); // <------ b assigned before a!
string a = Foobar();
string text = (a == null || !(a == "foo") || b == null || !(b == "bar")) ? ("Wait, what? I got (" + a + ", " + b + ")!") : "Yep, all good!";
string text2 = text;
string r = text2;
Console.WriteLine(r);
string Foobar()
{
    return s[i++];
}
```

It's also worth noting that a `switch` statement works as expected by printing `"Yep, all good!"`:

```c#
var s = new[] { "foo", "bar" };
var i = 0;
string Foobar() => s[i++];
string r;
switch (Foobar(), Foobar())
{
    case ("foo", "bar"): r =  "Yep, all good!"; break;
    case var (a, b)    : r = $"Wait, what? I got ({a}, {b})!"; break;
};
Console.WriteLine(r);
```

But imagine someone innocently refactors the above into a switch-expression and now they have introduced a subtle bug in their application.

This change is important enough that it needs to be called out and the effects demonstrated. Today, it's hidden deep in specification proposal notes.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a94537c5-754e-9dea-94a8-7573011f588c
* Version Independent ID: aba09118-22c5-2c6a-954a-b9659a9644ae
* Content: [What's new in C# 8.0 - C# Guide](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#switch-expressions)
* Content Source: [docs/csharp/whats-new/csharp-8.md](https://github.com/dotnet/docs/blob/master/docs/csharp/whats-new/csharp-8.md)
* Product: **dotnet-csharp**
* Technology: **csharp-whats-new**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

---

@atifaziz commented on [Tue Oct 29 2019](https://github.com/dotnet/docs/issues/15496#issuecomment-547363063)

Following is a simpler version of the demo without needing a local function with side-effects. Any object from the BCL with impure methods will do:

```c#
using var sr = new StringReader("foo\nbar");
var r = (sr.ReadLine(), sr.ReadLine()) switch
{
    ("foo", "bar") =>  "Yep, all good!",
    var (a, b)     => $"Wait, what? I got ({a}, {b})!",
};
Console.WriteLine(r); // prints: Wait, what? I got (bar, foo)!
```

---

@BillWagner commented on [Tue Oct 29 2019](https://github.com/dotnet/docs/issues/15496#issuecomment-547436682)

Thanks for bringing this to our attention @atifaziz 

I'm adding a couple folks on the C# LDM to discuss how we should explain this behavior.

/cc @gafter @MadsTorgersen 

---

@gafter commented on [Tue Oct 29 2019](https://github.com/dotnet/docs/issues/15496#issuecomment-547459405)

This is a bug.  The tuple expression is required to evaluate its operands in order, and in this case the compiler does not do so.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #39564

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs [hunk 1]
```diff
@@ -9409,6 +9409,193 @@ .maxstack  1
...
+        {
+            (""foo"", ""bar"") => ""Yep, all good!"",
+            var (a, b) => $""Wait, what? I got ({a}, {b})!"",
+        };
+        Console.WriteLine(r);
+    }
+}
+";
+            var compilation = CreateCompilation(source,
+                    options: TestOptions.DebugExe.WithOutputKind(OutputKind.ConsoleApplication))
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs]: >WithOutputKind(OutputKind.ConsoleApplication) [](start = 50, length = 45)

`DebugExe` already has this kind.

File: src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs [hunk 2]
```diff
@@ -9409,6 +9409,193 @@ .maxstack  1
...
+        IL_0085:  ldloc.0
+        IL_0086:  callvirt   ""void System.IDisposable.Dispose()""
+        IL_008b:  nop
+        IL_008c:  endfinally
+      }
+      IL_008d:  ret
+    }
+");
+            compilation = CreateCompilation(source,
+                    options: TestOptions.ReleaseExe.WithOutputKind(OutputKind.ConsoleApplication))
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/SwitchTests.cs]: >WithOutputKind(OutputKind.ConsoleApplication) [](start = 52, length = 45)

`ReleaseExe` already has this kind.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs [hunk 3]
```diff
@@ -504,14 +503,13 @@ bool usesOriginalInput(BoundDecisionDagNode node)
...
                     var fieldFetchEvaluation = new BoundDagFieldEvaluation(expr.Syntax, field, originalInput);
                     var temp = new BoundDagTemp(expr.Syntax, expr.Type, fieldFetchEvaluation);
-                    if (!tupleElementEvaluated[i])
-                    {
-                        storeToTemp(temp, expr);
-                    }
-
+                    Debug.Assert(!tupleElementEvaluated[i]);
+                    storeToTemp(temp, expr);
+                    tupleElementEvaluated[i] = true;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs]: >tupleElementEvaluated[i] = true; [](start = 20, length = 32)

does this still serve a purpose now? (it seems to only be used in assertions)

**Comment by atifaziz** [comment]
Thank you @gafter for fixing and @jcouv for reviewing.

**Pull request merged**